pytorch_lightning/core/optimizer.py

```diff
@@ -337,15 +337,26 @@ def _get_default_scheduler_config() -> Dict[str, Any]:
         "reduce_on_plateau": False,  # most often not ReduceLROnPlateau scheduler
         "monitor": None,  # value to monitor for ReduceLROnPlateau
         "strict": True,  # enforce that the monitor exists for ReduceLROnPlateau
-        "opt_idx": None,  # necessary to store opt_idx when optimizer frequencies are specified
+        "opt_idx": None,  # opt_idx assigned internally if not assigned by user
     }
 
 
-def _validate_scheduler_optimizer(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
-    if any(sch["scheduler"].optimizer not in optimizers for sch in lr_schedulers):
-        raise MisconfigurationException(
-            "Some schedulers are attached with an optimizer that wasn't returned from `configure_optimizers`."
-        )
+def _attach_scheduler_opt_idx(optimizers: List[Any], lr_schedulers: List[Any]) -> None:
+    for sch in lr_schedulers:
+        opt_found = sch["opt_idx"] is not None
+        if opt_found:
+            continue
+
+        for opt_idx, opt in enumerate(optimizers):
+            if sch["scheduler"].optimizer == opt:
+                sch["opt_idx"] = opt_idx
+                opt_found = True
+                break
+
+        if not opt_found:
+            raise MisconfigurationException(
+                "Some schedulers are attached with an optimizer that wasn't returned from `configure_optimizers`."
```

Comment:
I think this should differentiate between two different errors:

- The user provided an optimizer index but such optimizer does not exist
- The scheduler optimizer reference is not one of the existing optimizers

pseudocode:

```python
for sch in lr_schedulers:
    for opt_idx, opt in enumerate(optimizers):
        if sch["scheduler"].optimizer is opt:
            # assumes we keep the None default
            if sch["opt_idx"] is not None and sch["opt_idx"] != opt_idx:
                error: does not exist
            sch["opt_idx"] = opt_idx
            break
    else:
        error: the reference does not match
```

Reply:
ah! good catch